escrow/token_service.go

```diff
@@ -0,0 +1,159 @@
+//go:generate protoc -I . ./token_service.proto --go_out=plugins=grpc:.
+package escrow
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/singnet/snet-daemon/authutils"
+	"github.com/singnet/snet-daemon/blockchain"
+	"github.com/singnet/snet-daemon/token"
+	"golang.org/x/net/context"
+	"math/big"
+)
+
+type TokenService struct {
+	channelService          PaymentChannelService
+	prePaidUsageService     PrePaidService
+	tokenManager            token.Manager
+	validator               *ChannelPaymentValidator
+	serviceMetaData         blockchain.ServiceMetadata
+	allowedBlockNumberCheck func(blockNumber *big.Int) (err error)
+}
+
+type BlockChainDisabledTokenService struct {
+}
+
+func (service BlockChainDisabledTokenService) GetToken(ctx context.Context, request *TokenRequest) (reply *TokenReply, err error) {
+	return &TokenReply{}, nil
+}
+
+func NewTokenService(paymentChannelService PaymentChannelService,
+	usageService PrePaidService, tokenManager token.Manager, validator *ChannelPaymentValidator, metadata *blockchain.ServiceMetadata) *TokenService {
+
+	return &TokenService{
+		channelService:      paymentChannelService,
+		prePaidUsageService: usageService,
+		tokenManager:        tokenManager,
+		validator:           validator,
+		serviceMetaData:     *metadata,
+		allowedBlockNumberCheck: func(blockNumber *big.Int) error {
+			currentBlockNumber, err := validator.currentBlock()
+			if err != nil {
+				return err
+			}
+			differenceInBlockNumber := blockNumber.Sub(blockNumber, currentBlockNumber)
+			if differenceInBlockNumber.Abs(differenceInBlockNumber).Uint64() > authutils.AllowedBlockChainDifference {
+				return fmt.Errorf("authentication failed as the signature passed has expired")
+			}
+			return nil
+		},
+	}
+}
+
+func (service *TokenService) verifySignatureAndSignedAmountEligibility(channelId *big.Int,
+	latestAuthorizedAmount *big.Int, request *TokenRequest) (err error) {
+	channel, ok, err := service.channelService.PaymentChannel(&PaymentChannelKey{ID: channelId})
+
+	if !ok {
+		return fmt.Errorf("channel is not found, channelId: %v", channelId)
+	}
+	if err != nil {
+		return fmt.Errorf("error:%v was seen on retreiving details of channelID:%v",
+			err.Error(), channelId)
+	}
+	if channel.AuthorizedAmount.Cmp(latestAuthorizedAmount) > 0 {
```

Comment:
@anandrgitnirman , sorry, may be it is incorrect comment because I see in code below you allow `AuthorizedAmount` be equal to `latestAuthorizedAmount`. So the question is: what behavior do you expect?

Reply:
hi @vsbogd ,  if you need to renew a token , you dont need to send a Higer signed amount , you can just send the same amount as you last signed